### PR TITLE
tfsdk: Ensure other attribute plan modifiers are called after errors

### PR DESCRIPTION
**Pull request by bflad**
Reference: https://github.com/hashicorp/terraform-plugin-framework/issues/121

This is to prevent situations where warning diagnostics in other attributes are skipped based on when an error diagnostic might be non-deterministically found.

### CODE REVIEW

**Review by paddycarver — APPROVED** [review]

**Comment by kmoe** [comment]
Call an error in provider code that should halt execution of `PlanResourceChange` handling a _fatal error_: for example, the API client gets an expired token response or hits a rate limit, so it is not possible to complete the operation. 

The tradeoff here is now that the framework must take one of the following options:

1. The first fatal error encountered in any attribute plan modifier halts execution of plan modifiers. Since the order the attributes are visited in is currently random, this leads to inconsistent return values.

2. The first fatal error encountered in any plan modifier halts execution, and attributes are always visited in the same order, so the return value is deterministic.

3. Plan modifiers cannot throw fatal errors - error diags do not halt execution.

I believe this PR is proposing option 3, which matches the validation functionality, in which attribute validators cannot throw fatal errors. This seems fine - we just need to make sure we document what it means to return an error diag so users know what to expect.

None of these options is obvious in my opinion, so let's mention in the GoDoc for `Attribute.PlanModifiers` whether the user should expect other plan modifiers to run if one errors.

**Comment by bflad** [comment]
That is an excellent write-up, @kmoe! Let's sync up later today to discuss this further.

**Comment by paddycarver** [comment]
It seems to me it's worth calling out the difference between attribute-level and resource-level fatal.

The current implementation has a _resource-level_ fatal. It stops not only plan modification for that attribute, but for all other attributes.

The proposed implementation has an _attribute-level_ fatal. It stops plan modification for that attribute, but other attributes continue to have _their_ plan modification run.

The only time I can think of that the proposed implementation would have unexpected consequences for provider developers is when a nested attribute's plan customization did not fully complete before its parent's plan customization is run. Can anyone else think of scenarios in which the proposed behavior would be surprising?

**Comment by kmoe** [comment]
Good distinction. Putting it this way makes it even clearer to me that the right approach is to make attribute plan modifier errors attribute-level fatal.

I can't think of a situation in which this behaviour would be surprising or unwelcome as long as it's clearly documented. Happy to approve this PR with something added to the GoDoc for this.

File: tfsdk/attribute.go [hunk 1]
```diff
@@ -75,10 +75,16 @@ type Attribute struct {
...
+	// plan time. Attribute-level plan modifications occur before any
+	// resource-level plan modifications.
+	//
+	// Any errors will prevent further execution of this sequence
+	// of modifiers and modifiers associated with nested Attribute, but have no
+	// effect with other Attribute in the Schema.
+	//
+	// Plan modification only applies to resources, not data sources or
+	// providers. Setting PlanModifiers on a data source or provider attribute
+	// will have no effect.
```
Comment (bflad) [inline tfsdk/attribute.go:87]: I must admit that this makes me 🤔  about `Attribute` being an interface and having `ProviderAttribute`, `ResourceAttribute`, and `DataSourceAttribute` to distinguish available functionality between these at compiler-time.

File: tfsdk/attribute.go [hunk 2]
```diff
@@ -75,10 +75,16 @@ type Attribute struct {
...
 	// PlanModifiers defines a sequence of modifiers for this attribute at
-	// plan time.
-	// Please note that plan modification only applies to resources, not
-	// data sources. Setting PlanModifiers on a data source attribute will
-	// have no effect.
+	// plan time. Attribute-level plan modifications occur before any
+	// resource-level plan modifications.
+	//
+	// Any errors will prevent further execution of this sequence
+	// of modifiers and modifiers associated with nested Attribute, but have no
```
Comment (kmoe) [inline tfsdk/attribute.go:82]:
```suggestion
	// of modifiers and modifiers associated with any nested Attribute, but will not
```

File: tfsdk/attribute.go [hunk 3]
```diff
@@ -75,10 +75,16 @@ type Attribute struct {
...
+	// effect with other Attribute in the Schema.
```
Comment (kmoe) [inline tfsdk/attribute.go:83]:
```suggestion
	// prevent execution of PlanModifiers on any other Attribute in the Schema.
```

File: tfsdk/attribute.go (hunk 1 above)
Comment (kmoe) [reply tfsdk/attribute.go:87]: Sounds like a good idea. I agree it's confusing having a field on the `Attribute` struct that sometimes does nothing.

**Review by kmoe — APPROVED** [review]
LGTM with some suggestions for wording

**Comment by bflad** [comment]
It looks like `(Attribute).modifyPlan()` also needed to be taught to ignore previous errors to ensure attribute-level plan modifiers were not skipped. To ensure the expected behaviors of that method, added exhaustive unit testing to cover all the branching logic.

File: tfsdk/attribute.go (hunk 1 above)
Comment (bflad) [reply tfsdk/attribute.go:86]: Created https://github.com/hashicorp/terraform-plugin-framework/issues/132

**Review by paddycarver — APPROVED** [review]

**Pull request merged**